### PR TITLE
Adding the HLD of Object Index

### PR DESCRIPTION
**Pull request by VenkyOS**
Signed-off-by: Venkataraman Padmanabhan <venkataraman.padmanabhan@seagate.com>

### CODE REVIEW

File: doc/HLD_of_Motr_Object_Index.rst [hunk 1]
```diff
@@ -0,0 +1,338 @@
+=========================================
+High level design of a Motr Object Index
```
Comment (huanghua78) [inline doc/HLD_of_Motr_Object_Index.rst:2]: s
/High level design of a Motr Object Index
/High Level Design of a Motr Object Index

File: doc/HLD_of_Motr_Object_Index.rst [hunk 2]
```diff
@@ -0,0 +1,338 @@
...
+
+This document presents a high level design (HLD) of an Object Index for Motr M0 core. The main purposes of this document are: (i) to be inspected by M0 architects and peer designers to ascertain that high level design is aligned with M0 architecture and other designs, and contains no defects, (ii) to be a source of material for Active Reviews of Intermediate Design (ARID) and detailed level design (DLD) of the same component, (iii) to serve as a design reference document.
+
+The intended audience of this document consists of M0 customers, architects, designers and developers.
+
+**************
+Introduction
+**************
+
+The object index performs the function of a metadata layer on top of M0 storage objects. A M0 storage object (stob) is a flat address space where one can read from or write on, with block size granularity. Stobs have no metadata associated with them. To be used as files (or components of files, aka stripes), additional metadata must be associated with the stobs.
```
Comment (huanghua78) [inline doc/HLD_of_Motr_Object_Index.rst:13]: s
/where one can read from or write on
where one can read from or write to

File: doc/HLD_of_Motr_Object_Index.rst (hunk 1 above)
Comment (huanghua78) [reply doc/HLD_of_Motr_Object_Index.rst:2]: Let's add a "OBSOLETED" at the end of this title.
Some of the detailed implementation are out-dated. But the basic idea is still valid.

Comment (VenkyOS) [reply doc/HLD_of_Motr_Object_Index.rst:2]: Comment Incorporated.

File: doc/HLD_of_Motr_Object_Index.rst (hunk 2 above)
Comment (VenkyOS) [reply doc/HLD_of_Motr_Object_Index.rst]: Comments Incorporated.

**Review by huanghua78 — APPROVED** [review]

File: doc/HLD_of_Motr_Object_Index.rst [hunk 3]
```diff
@@ -0,0 +1,373 @@
...
+
+                    struct fileattr_basic_rec *co_fab_rec;
+
+                    struct object_index_rec *co_oi_rec; /* pfid, filename */
+
+    };
+
+The `*_rec` members are pointers to the records from the database tables. These records may or may not be populated at various stages in cob life.
+
+The co_stob reference is also likely to remain unset, as metadata operations will not frequently affect the underlying storage object (and, indeed, the storage object is likely to live on a different node).
```
[the remark this answers is not in the document]
Comment (VenkyOS) [reply doc/HLD_of_Motr_Object_Index.rst]: Incorporated.

File: doc/HLD_of_Motr_Object_Index.rst [hunk 4]
```diff
@@ -0,0 +1,373 @@
...
+
+- namespace table
+
+  - key is {parent_fid, filename}
+
+  - record is {child_fid, nlink, attrs}
+
+  - if nlink > 0, attrs = {size, mactime, omg_ref, nlink}; else attrs = {}
+
+  - multiple keys may point to the same record for hardlinks, if the database can support this. Otherwise, we store the attrs in one of the records only (link number 0). (This    leads to a long sequence of operations to delete a hardlink, but straightforward.)
```
[the remark this answers is not in the document]
Comment (VenkyOS) [reply doc/HLD_of_Motr_Object_Index.rst]: Incorporated.

File: doc/HLD_of_Motr_Object_Index.rst [hunk 5]
```diff
@@ -0,0 +1,373 @@
...
+
+- A global object (gob) is an object describing a striped file, by referring to a collection of component objects.
+
+**************
+Requirements
+**************
+
+- [R.M0.BACK-END.OBJECT-INDEX]: an object index allows back-end to locate an object by its fid
+
+- [R.M0.BACK-END.INDEXING]: back-end has mechanisms to build meta-data indices
```
[the remark this answers is not in the document]
Comment (VenkyOS) [reply doc/HLD_of_Motr_Object_Index.rst]: Incorporated.

File: doc/HLD_of_Motr_Object_Index.rst [hunk 6]
```diff
@@ -0,0 +1,373 @@
+=========================================
+High level design of a Motr Object Index
+=========================================
+
+This document presents a High Level Design (HLD) of an Object Index for Motr M0 core. The main purposes of this document are: (i) to be inspected by M0 architects and peer designers to ascertain that high level design is aligned with M0 architecture and other designs, and contains no defects, (ii) to be a source of material for Active Reviews of Intermediate Design (ARID) and detailed level design (DLD) of the same component, (iii) to serve as a design reference document.
+
+The intended audience of this document consists of M0 customers, architects, designers and developers.
```
[the remark this answers is not in the document]
Comment (VenkyOS) [reply doc/HLD_of_Motr_Object_Index.rst:7]: Deferred.

File: doc/HLD_of_Motr_Object_Index.rst [hunk 7]
```diff
@@ -0,0 +1,373 @@
+=========================================
+High level design of a Motr Object Index
```
[the remark this answers is not in the document]
Comment (VenkyOS) [reply doc/HLD_of_Motr_Object_Index.rst]: Incorporated.

File: doc/HLD_of_Motr_Object_Index.rst [hunk 8]
```diff
@@ -0,0 +1,373 @@
...
+=========================================
+
+This document presents a High Level Design (HLD) of an Object Index for Motr M0 core. The main purposes of this document are: (i) to be inspected by M0 architects and peer designers to ascertain that high level design is aligned with M0 architecture and other designs, and contains no defects, (ii) to be a source of material for Active Reviews of Intermediate Design (ARID) and detailed level design (DLD) of the same component, (iii) to serve as a design reference document.
```
[the remark this answers is not in the document]
Comment (VenkyOS) [reply doc/HLD_of_Motr_Object_Index.rst]: Incorporated.

File: doc/HLD_of_Motr_Object_Index.rst [hunk 9]
```diff
@@ -0,0 +1,373 @@
...
+
+The intended audience of this document consists of M0 customers, architects, designers and developers.
+
+**************
+Introduction
+**************
+
+The object index performs the function of a metadata layer on top of M0 storage objects. A M0 storage object (stob) is a flat address space where one can read from or write to, with block size granularity. Stobs have no metadata associated with them. To be used as files (or components of files, aka stripes), additional metadata must be associated with the stobs.
```
[the remark this answers is not in the document]
Comment (VenkyOS) [reply doc/HLD_of_Motr_Object_Index.rst]: Incorporated.

**Pull request merged**